internal/servers/worker/status.go

```diff
@@ -115,6 +116,35 @@ func (w *Worker) sendWorkerStatus(cancelCtx context.Context) {
 	})
 	if err != nil {
 		w.logger.Error("error making status request to controller", "error", err)
+		// Check for last successful status. Ignore nil last status, this probably
+		// means that we've never connected to a controller, and as such probably
+		// don't have any sessions to worry about anyway.
+		//
+		// If a length of time has passed since we've been able to communicate, we
+		// want to start terminating all sessions as a "break glass" kind of
+		// scenario, as there will be no way we can really tell if these
+		// connections should continue to exist.
+		lastStatus := w.LastStatusSuccess()
+		if lastStatus != nil && lastStatus.StatusTime.Before(time.Now().Add(statusFailGrace*-1)) {
+			w.logger.Warn("status error grace period has expired, canceling all sessions on worker",
+				"last_status_time", lastStatus.StatusTime.String(),
+				"grace_period", statusFailGrace,
+			)
+
+			// Iterate through all sessions. Transition any sessions with state
+			// <SESSIONSTATUS_SESSIONSTATUS_CANCELING to that state.
+			w.sessionInfoMap.Range(func(sessionId, siRaw interface{}) bool {
+				si := siRaw.(*sessionInfo)
+				if si.status < pbs.SESSIONSTATUS_SESSIONSTATUS_CANCELING {
+					w.logger.Warn("canceling session due to worker communication failure", "session_id", sessionId)
+					si.Lock()
+					si.status = pbs.SESSIONSTATUS_SESSIONSTATUS_CANCELING
```

Comment:
I realize that setting it to canceling here allows the logic below to end up tearing down the session, but the issue here is that the session isn't necessarily canceled -- so if the status starts being sent/received properly again, we'd have a mismatch on session state from the controller.

Instead, you should terminate and mark closed the various connections within the sessions. Users won't be able to create new connections because if the communication with the controller is down it can't be authorized, and if the communication is up and the session was in fact canceled then it also won't be authorized.

Reply:
I was assuming that the controller would advance the status appropriately, but it seems not to do so and there's other connection-specific logic that may need to be followed. Will update accordingly.

Reply:
Some notes for myself and anyone else looking:

My assumption, from a purely status perspective, was that the worker could advance the state of the session appropriately once the connection came back online (_if_ it came back online in time, before it got reaped), but that does not seem to be the case. Based on this, purely closing the connections should be fine.

The question then is, how does the _connection's_ state get updated properly? Looks like we clean connections appropriately in the controller's [`Status`](https://github.com/hashicorp/boundary/blob/e443bd4bb448f0351096503d2388fe2e80102064/internal/servers/controller/handlers/workers/worker_service.go#L54-L166) handler, but the connection will still be accounted for if the status is not properly advanced in the worker. Nothing seems to do so at this time; synchronizing the connection status on close seems to be [dependent on the response from the controller](https://github.com/hashicorp/boundary/blob/e443bd4bb448f0351096503d2388fe2e80102064/internal/servers/worker/session.go#L288-L294).

The logical answer seems to be advancing the connection state if it can't be properly advanced via the controller.

Reply:
I still wonder if this is the right approach; unless the session should be able to be re-used after we've terminated all connections on it, this _seems_ like the cleanest approach - advancing the state allows already established tear-down logic to take place and for the session to eventually be removed on the worker.

All that seems to be necessary here is:

* Ensuring that the status gets advanced [here](https://github.com/hashicorp/boundary/blob/e443bd4bb448f0351096503d2388fe2e80102064/internal/servers/controller/handlers/workers/worker_service.go#L110-L115) on appropriate update. 
* Dead connection cleanup will handle canceling the connections as appropriate.
* Cleaning up _orphaned sessions_ if they are not reported by the worker (this seems to be something that could possibly be an issue, independent of any work here currently).

For completeness, we might also want to manually force connection states on the worker if [`CloseConnection`](https://github.com/hashicorp/boundary/blob/e443bd4bb448f0351096503d2388fe2e80102064/internal/servers/worker/session.go#L243) fails too, but it does not seem 100% necessary as the worker is likely to reap it from the session on the next `Status` call.

Reply:
The problem is that sessions can live cross-worker; connections for a session can be handled by any worker so long as the client provides the right TOFU token. So unless a client has specifically indicated to a worker that it wants the session canceled, all a worker can know is that its connections related to a session should be closed. There may be other workers still connected to controllers handling other connections for that session that aren't hitting this failure condition.

Reply:
:+1: gotcha; this makes sense. Will update.